third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,1161 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! AST Simplifier
+//!
+//! Simplify the AST before conversion to bytecode.
+//! - flow-insensitive constant propagation and folding
+//! - simple expression simplification
+//!
+//! Preconditions:
+//! - have previously run `check_for_unused_vars_and_params` pass to warn about unused vars,
+//!   which may be eliminated here.
+//!
+//! More details:
+//! - Do flow-insensitive constant propagation:
+//!   - identify "possibly modified" symbols which may have more than one value
+//!   - for safe symbols whose value is a constant value, propagate
+//!     the value to the use site to enable simplifying code:
+//!     - inline a constant value
+//!     - flag uses of uninitialized and unassigned variables
```

Comment:
It's trivial to recognize the flow-insensitive case here: a variable which is declared in a let with no binding and is not modified.  I suppose we could remove it if `uninitialized_use_checker` is jealous of having those easy cases, but it's not a duplication because it's an error (not a warning) and compilation will not reach `uninitialized_use_checker`.